File: go/zebra-puzzle/zebra_puzzle.go
package zebra

// Define constants for properties
type Color int

const (
	Red Color = iota
	Green
	Ivory
	Yellow
	Blue
)

var Colors = []Color{Red, Green, Ivory, Yellow, Blue}

func (c Color) String() string { return []string{"Red", "Green", "Ivory", "Yellow", "Blue"}[c] }

type Nationality int

const (
	Englishman Nationality = iota
	Spaniard
	Ukrainian
	Norwegian
	Japanese
)

var Nationalities = []Nationality{Englishman, Spaniard, Ukrainian, Norwegian, Japanese}

func (n Nationality) String() string {
	return []string{"Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"}[n]
}

type Pet int

const (
	Dog Pet = iota
	Snails
	Fox
	Horse
	Zebra
)

var Pets = []Pet{Dog, Snails, Fox, Horse, Zebra}

func (p Pet) String() string { return []string{"Dog", "Snails", "Fox", "Horse", "Zebra"}[p] }

type Drink int

const (
	Coffee Drink = iota
	Tea
	Milk
	OrangeJuice
	Water
)

var Drinks = []Drink{Coffee, Tea, Milk, OrangeJuice, Water}

func (d Drink) String() string { return []string{"Coffee", "Tea", "Milk", "Orange Juice", "Water"}[d] }

type Cigarette int

const (
	OldGold Cigarette = iota
	Kools
	Chesterfields
	LuckyStrike
	Parliaments
)

var Cigarettes = []Cigarette{OldGold, Kools, Chesterfields, LuckyStrike, Parliaments}

func (c Cigarette) String() string {
	return []string{"Old Gold", "Kools", "Chesterfields", "Lucky Strike", "Parliaments"}[c]
}

// House structure
type House struct {
	Position    int // 0-4
	Color       Color
	Nationality Nationality
	Pet         Pet
	Drink       Drink
	Cigarette   Cigarette
}

type Solution struct {
	DrinksWater string
	OwnsZebra   string
}

// permutations generates all permutations of integers from 0 to n-1.
// Uses Heap's algorithm.
func permutations(n int) [][]int {
	p := make([]int, n)
	for i := 0; i < n; i++ {
		p[i] = i
	}

	var result [][]int
	var generate func(k int, arr []int)

	generate = func(k int, arr []int) {
		if k == 1 {
			perm := make([]int, n)
			copy(perm, arr)
			result = append(result, perm)
			return
		}

		// Generate permutations for k-1 recursively
		generate(k-1, arr)

		// Generate permutations for k by swapping the (k-1)th element
		for i := 0; i < k-1; i++ {
			// Swap based on whether k is odd or even
			if k%2 == 0 {
				arr[i], arr[k-1] = arr[k-1], arr[i]
			} else {
				arr[0], arr[k-1] = arr[k-1], arr[0]
			}
			generate(k-1, arr)
		}
	}

	generate(n, p)
	return result
}

// abs helper function
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// Helper function to find house by property
func findHouseByNat(houses []House, nat Nationality) *House {
	for i := range houses {
		if houses[i].Nationality == nat {
			return &houses[i]
		}
	}
	return nil
}
func findHouseByColor(houses []House, col Color) *House {
	for i := range houses {
		if houses[i].Color == col {
			return &houses[i]
		}
	}
	return nil
}
func findHouseByPet(houses []House, pet Pet) *House {
	for i := range houses {
		if houses[i].Pet == pet {
			return &houses[i]
		}
	}
	return nil
}
// func findHouseByDrink(houses []House, dri Drink) *House {
// 	for i := range houses { if houses[i].Drink == dri { return &houses[i] } }
// 	return nil
// }
func findHouseByCig(houses []House, cig Cigarette) *House {
	for i := range houses {
		if houses[i].Cigarette == cig {
			return &houses[i]
		}
	}
	return nil
}

// Helper function to find position by property
// func findPosByNat(houses []House, nat Nationality) int {
// 	for i := range houses { if houses[i].Nationality == nat { return i } }
// 	return -1
// }
func findPosByColor(houses []House, col Color) int {
	for i := range houses {
		if houses[i].Color == col {
			return i
		}
	}
	return -1
}
func findPosByPet(houses []House, pet Pet) int {
	for i := range houses {
		if houses[i].Pet == pet {
			return i
		}
	}
	return -1
}
// func findPosByDrink(houses []House, dri Drink) int {
// 	for i := range houses { if houses[i].Drink == dri { return i } }
// 	return -1
// }
func findPosByCig(houses []House, cig Cigarette) int {
	for i := range houses {
		if houses[i].Cigarette == cig {
			return i
		}
	}
	return -1
}

// Function to check all constraints for a given house configuration
func checkConstraints(houses []House) bool {
	// Constraints checked earlier by pruning: 6, 9, 10, 15

	// 2. The Englishman lives in the red house.
	hEng := findHouseByNat(houses, Englishman)
	if hEng == nil || hEng.Color != Red {
		return false
	}

	// 3. The Spaniard owns the dog.
	hSpa := findHouseByNat(houses, Spaniard)
	if hSpa == nil || hSpa.Pet != Dog {
		return false
	}

	// 4. Coffee is drunk in the green house.
	hGreen := findHouseByColor(houses, Green)
	if hGreen == nil || hGreen.Drink != Coffee {
		return false
	}

	// 5. The Ukrainian drinks tea.
	hUkr := findHouseByNat(houses, Ukrainian)
	if hUkr == nil || hUkr.Drink != Tea {
		return false
	}

	// 7. The Old Gold smoker owns snails.
	hOG := findHouseByCig(houses, OldGold)
	if hOG == nil || hOG.Pet != Snails {
		return false
	}

	// 8. Kools are smoked in the yellow house.
	hYellow := findHouseByColor(houses, Yellow)
	if hYellow == nil || hYellow.Cigarette != Kools {
		return false
	}

	// 11. The man who smokes Chesterfields lives in the house next to the man with the fox.
	chesterPos := findPosByCig(houses, Chesterfields)
	foxPos := findPosByPet(houses, Fox)
	if chesterPos == -1 || foxPos == -1 || abs(chesterPos-foxPos) != 1 {
		return false
	}

	// 12. Kools are smoked in the house next to the house where the horse is kept.
	koolsPos := findPosByCig(houses, Kools) // Already found via hYellow
	horsePos := findPosByPet(houses, Horse)
	if koolsPos == -1 || horsePos == -1 || abs(koolsPos-horsePos) != 1 {
		return false
	}

	// 13. The Lucky Strike smoker drinks orange juice.
	hLS := findHouseByCig(houses, LuckyStrike)
	if hLS == nil || hLS.Drink != OrangeJuice {
		return false
	}

	// 14. The Japanese smokes Parliaments.
	hJap := findHouseByNat(houses, Japanese)
	if hJap == nil || hJap.Cigarette != Parliaments {
		return false
	}

	// All constraints passed
	return true
}

func SolvePuzzle() Solution {
	perms := permutations(5) // Generate permutations of [0, 1, 2, 3, 4]

	// Iterate through all possible assignments (permutations) for each category
	for _, natPerm := range perms { // Nationality permutation
		houses := make([]House, 5)
		for i := 0; i < 5; i++ {
			houses[i].Position = i
			// Assign nationality based on the current permutation
			houses[i].Nationality = Nationality(natPerm[i])
		}

		// Constraint 10: The Norwegian lives in the first house (index 0).
		if houses[0].Nationality != Norwegian {
			continue
		}

		for _, colPerm := range perms { // Color permutation
			for i := 0; i < 5; i++ {
				houses[i].Color = Color(colPerm[i])
			}

			// Constraint 15: The Norwegian lives next to the blue house.
			// Since Norwegian is in house 0, Blue must be in house 1 (index 1).
			if houses[1].Color != Blue {
				continue
			}

			// Constraint 6: The green house is immediately to the right of the ivory house.
			ivoryPos, greenPos := -1, -1
			for i := 0; i < 5; i++ {
				if houses[i].Color == Ivory {
					ivoryPos = i
				}
				if houses[i].Color == Green {
					greenPos = i
				}
			}
			// Check if found and if green is right of ivory
			if ivoryPos == -1 || greenPos == -1 || greenPos != ivoryPos+1 {
				continue
			}

			for _, petPerm := range perms { // Pet permutation
				for i := 0; i < 5; i++ {
					houses[i].Pet = Pet(petPerm[i])
				}

				for _, driPerm := range perms { // Drink permutation
					for i := 0; i < 5; i++ {
						houses[i].Drink = Drink(driPerm[i])
					}

					// Constraint 9: Milk is drunk in the middle house (index 2).
					if houses[2].Drink != Milk {
						continue
					}

					for _, cigPerm := range perms { // Cigarette permutation
						for i := 0; i < 5; i++ {
							houses[i].Cigarette = Cigarette(cigPerm[i])
						}

						// Check all remaining constraints
						if checkConstraints(houses) {
							// Found the solution
							waterDrinker := ""
							zebraOwner := ""
							for _, h := range houses {
								if h.Drink == Water {
									waterDrinker = h.Nationality.String()
								}
								if h.Pet == Zebra {
									zebraOwner = h.Nationality.String()
								}
							}
							// Ensure both were found before returning
							if waterDrinker != "" && zebraOwner != "" {
								return Solution{DrinksWater: waterDrinker, OwnsZebra: zebraOwner}
							}
							// If only one is found, something is wrong with the logic or constraints
						}
					}
				}
			}
		}
	}

	// Should not happen if a solution exists and logic is correct
	panic("Solution not found")
}

File: java/zebra-puzzle/src/main/java/ZebraPuzzle.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class ZebraPuzzle {

    // --- Enums for Categories ---
    enum Color { RED, GREEN, IVORY, YELLOW, BLUE }
    enum Nationality { ENGLISHMAN, SPANIARD, UKRAINIAN, NORWEGIAN, JAPANESE }
    enum Pet { DOG, SNAILS, FOX, HORSE, ZEBRA }
    enum Drink { COFFEE, TEA, MILK, ORANGE_JUICE, WATER }
    enum Hobby { DANCING, PAINTER, READING, FOOTBALL, CHESS }

    // --- Solution Holder ---
    private static final Solution SOLVED_PUZZLE = solve();

    // --- House Representation (Implicit: Index 0-4 represents house number) ---
    private static class Solution {
        final Color[] colors;
        final Nationality[] nationalities;
        final Pet[] pets;
        final Drink[] drinks;
        final Hobby[] hobbies;

        Solution(Color[] c, Nationality[] n, Pet[] p, Drink[] d, Hobby[] h) {
            colors = c; nationalities = n; pets = p; drinks = d; hobbies = h;
        }

        // Helper to get index of a specific attribute
        int indexOf(Object attribute) {
            if (attribute instanceof Color) return indexOf(colors, (Color) attribute);
            if (attribute instanceof Nationality) return indexOf(nationalities, (Nationality) attribute);
            if (attribute instanceof Pet) return indexOf(pets, (Pet) attribute);
            if (attribute instanceof Drink) return indexOf(drinks, (Drink) attribute);
            if (attribute instanceof Hobby) return indexOf(hobbies, (Hobby) attribute);
            return -1;
        }

        private <T> int indexOf(T[] array, T value) {
            for (int i = 0; i < array.length; i++) {
                if (array[i] == value) return i;
            }
            return -1;
        }

        // Helper to check if house at index i has attribute value
        boolean has(int index, Object attribute) {
             if (attribute instanceof Color) return colors[index] == attribute;
             if (attribute instanceof Nationality) return nationalities[index] == attribute;
             if (attribute instanceof Pet) return pets[index] == attribute;
             if (attribute instanceof Drink) return drinks[index] == attribute;
             if (attribute instanceof Hobby) return hobbies[index] == attribute;
             return false;
        }
    }

    // --- Solver Logic ---
    private static Solution solve() {
        // Generate all permutations for each category
        List<Color[]> colorPermutations = generatePermutations(Color.values());
        List<Nationality[]> nationalityPermutations = generatePermutations(Nationality.values());
        List<Pet[]> petPermutations = generatePermutations(Pet.values());
        List<Drink[]> drinkPermutations = generatePermutations(Drink.values());
        List<Hobby[]> hobbyPermutations = generatePermutations(Hobby.values());

        // Iterate through all combinations of permutations
        for (Color[] colors : colorPermutations) {
            // Constraint 10: The Norwegian lives in the first house (house 0).
            // Constraint 15: The Norwegian lives next to the blue house.
            // Apply these early to prune possibilities for nationality and color.
            if (colors[1] != Color.BLUE) continue; // If house 0 is Norwegian, house 1 must be blue

            for (Nationality[] nationalities : nationalityPermutations) {
                 if (nationalities[0] != Nationality.NORWEGIAN) continue; // Constraint 10

                for (Pet[] pets : petPermutations) {
                    for (Drink[] drinks : drinkPermutations) {
                         // Constraint 9: The person in the middle house (house 2) drinks milk.
                         if (drinks[2] != Drink.MILK) continue;

                        for (Hobby[] hobbies : hobbyPermutations) {
                            Solution potentialSolution = new Solution(colors, nationalities, pets, drinks, hobbies);
                            if (checkConstraints(potentialSolution)) {
                                return potentialSolution; // Found the unique solution
                            }
                        }
                    }
                }
            }
        }
        throw new IllegalStateException("Solution not found!"); // Should not happen if puzzle is solvable
    }

    // --- Constraint Checking ---
    private static boolean checkConstraints(Solution s) {
        // 1. There are five houses. (Implicit by array size 5)

        // 2. The Englishman lives in the red house.
        if (s.colors[s.indexOf(Nationality.ENGLISHMAN)] != Color.RED) return false;

        // 3. The Spaniard owns the dog.
        if (s.pets[s.indexOf(Nationality.SPANIARD)] != Pet.DOG) return false;

        // 4. The person in the green house drinks coffee.
        if (s.drinks[s.indexOf(Color.GREEN)] != Drink.COFFEE) return false;

        // 5. The Ukrainian drinks tea.
        if (s.drinks[s.indexOf(Nationality.UKRAINIAN)] != Drink.TEA) return false;

        // 6. The green house is immediately to the right of the ivory house.
        int greenIdx = s.indexOf(Color.GREEN);
        int ivoryIdx = s.indexOf(Color.IVORY);
        if (greenIdx != ivoryIdx + 1) return false;

        // 7. The snail owner likes to go dancing.
        if (s.hobbies[s.indexOf(Pet.SNAILS)] != Hobby.DANCING) return false;

        // 8. The person in the yellow house is a painter.
        if (s.hobbies[s.indexOf(Color.YELLOW)] != Hobby.PAINTER) return false;

        // 9. The person in the middle house drinks milk. (Checked earlier during generation)
        // if (s.drinks[2] != Drink.MILK) return false; // Redundant check

        // 10. The Norwegian lives in the first house. (Checked earlier during generation)
        // if (s.nationalities[0] != Nationality.NORWEGIAN) return false; // Redundant check

        // 11. The person who enjoys reading lives in the house next to the person with the fox.
        int readingIdx = s.indexOf(Hobby.READING);
        int foxIdx = s.indexOf(Pet.FOX);
        if (Math.abs(readingIdx - foxIdx) != 1) return false;

        // 12. The painter's house is next to the house with the horse.
        int painterIdx = s.indexOf(Hobby.PAINTER);
        int horseIdx = s.indexOf(Pet.HORSE);
        if (Math.abs(painterIdx - horseIdx) != 1) return false;

        // 13. The person who plays football drinks orange juice.
        if (s.drinks[s.indexOf(Hobby.FOOTBALL)] != Drink.ORANGE_JUICE) return false;

        // 14. The Japanese person plays chess.
        if (s.hobbies[s.indexOf(Nationality.JAPANESE)] != Hobby.CHESS) return false;

        // 15. The Norwegian lives next to the blue house. (Partially checked earlier)
        int norwegianIdx = s.indexOf(Nationality.NORWEGIAN); // Should be 0
        int blueIdx = s.indexOf(Color.BLUE);
        if (Math.abs(norwegianIdx - blueIdx) != 1) return false;

        // All constraints passed
        return true;
    }


    // --- Permutation Generation (Heap's Algorithm) ---
    private static <T> List<T[]> generatePermutations(T[] values) {
        List<T[]> list = new java.util.ArrayList<>();
        generate(values.length, values.clone(), list);
        return list;
    }

    private static <T> void generate(int k, T[] a, List<T[]> output) {
        if (k == 1) {
            output.add(a.clone());
        } else {
            generate(k - 1, a, output);
            for (int i = 0; i < k - 1; i++) {
                if (k % 2 == 0) {
                    swap(a, i, k - 1);
                } else {
                    swap(a, 0, k - 1);
                }
                generate(k - 1, a, output);
            }
        }
    }

    private static <T> void swap(T[] a, int i, int j) {
        T temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    private String toTitleCase(String input) {
        if (input == null || input.isEmpty()) {
            return input;
        }
        return input.substring(0, 1).toUpperCase() + input.substring(1).toLowerCase();
    }

    // --- Public API Methods ---
    String getWaterDrinker() {
        int waterDrinkerIndex = SOLVED_PUZZLE.indexOf(Drink.WATER);
        if (waterDrinkerIndex == -1) {
             throw new IllegalStateException("Could not find water drinker in the solution.");
        }
        return toTitleCase(SOLVED_PUZZLE.nationalities[waterDrinkerIndex].name());
    }

    String getZebraOwner() {
        int zebraOwnerIndex = SOLVED_PUZZLE.indexOf(Pet.ZEBRA);
         if (zebraOwnerIndex == -1) {
             throw new IllegalStateException("Could not find zebra owner in the solution.");
        }
        return toTitleCase(SOLVED_PUZZLE.nationalities[zebraOwnerIndex].name());
    }
}
